### ISSUE TITLE
Screen brightness

### ISSUE THREAD

**Issue by tuftedocelot** [Lokaltog/candybar#38]
Screen brightness would be a useful widget to have especially for laptop users. I found some GNOME code as for how they use dbus to control brightness. My C is terrible (practically nonexistent), but I'll take a stab at it later. For reference if anyone wants to take a crack at it: https://git.gnome.org/browse/gnome-settings-daemon/commit/?id=a4c3299c97ced9c9bcdc69800464103ba47a0a56
##

**Comment by Lokaltog** [comment]
A screen brightness indicator could be very useful! And don't worry about your C skills, this is a learning experiment for me too, and it's awesome if you can contribute _something_, even if it's only a basic proof of concept!

If you decide to try to implement it, check out some of the other widgets that use dbus (battery and notifications). Also see the [loadable modules](https://github.com/Lokaltog/wkline/compare/feature/loadable-modules) branch, the widget format has been changed slightly from what's currently in the develop branch.

**Comment by tuftedocelot** [comment]
I think sysfs might be the easier way to read the brightness from what I can tell

**Comment by Lokaltog** [comment]
It may be easier, but I think we should avoid reading from the file system in this case. With dbus it should be possible to wait for changes instead of fetching them on an interval (similar to e.g. the volume widget), I think @ricardo-vieira found a guide that demonstrates this for battery information.

**Comment by tuftedocelot** [comment]
Hmm, so here's what I've found so far (based on Googling how dbus works)...I can run dbus-monitor with --system, adjust the backlight, and I get the PropertiesChanged output, but none of the output actually seems to contain the new value. Example: https://gist.github.com/tuftedocelot/9435298 I can get the path of the device, but not what is the change of the value.

**Comment by tuftedocelot** [comment]
udev seems useful to actually get the backlight brightness value. User adjusts backlight > dbus registers the change > pass the device name to udev > use udev to read the new value

**Comment by Lokaltog** [comment]
Do you have a code example for that?

**Comment by tuftedocelot** [comment]
Not code, no, just bits of commands that I haven't put together yet. Running `dbus-monitor --system` and changing the brightness gives the results in the gist above. If you grab line 8 of the gist and run `udevadm info -a -p <line8>`, you should get a line like `ATTR{brightness}=="8"`

**Comment by Lokaltog** [comment]
I think we should avoid reading files and running shell commands if it's possible, so if udev provides a low-level library to read the udev brightness data that would be better.

**Comment by tuftedocelot** [comment]
Right, I agree and that's the plan I'd like to follow. This has just been what I've been toying with so far.

**Comment by Lokaltog** [comment]
Ah, I see. Great work, I'm looking forward to seeing the results!

**Comment by tuftedocelot** [comment]
Meanwhile, eight months later...I came back to using candybar and looked into interfacing with C and udev to get screen brightness and write a widget. Based on http://www.signal11.us/oss/udev/, I made a small demo application which grabs the needed values based on events. I tried making that into a widget and while it compiles fine, it fails to load (unsurprisingly). I suspect the model widget to follow is volume (since it's also based around events), but once again, I've exhausted my C knowledge and I'm not sure how to handle events with the polling (I think?) in the widget.

https://github.com/tuftedocelot/candybar/tree/udev
